pkg/rancher-desktop/main/diagnostics/wslFromStore.ts

```diff
@@ -0,0 +1,221 @@
+      ['query', 'HKLM\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion', '/v', 'UBR', '/t', 'REG_DWORD'],
+      { stdio: ['ignore', 'pipe', console] });
+    const lines = stdout.split(/\n/).map(line => line.trim().split(/\s+/));
+    const patchString = lines.find(([key]) => key === 'UBR')?.pop() ?? '0x0';
+    const patch = parseInt(patchString, 16);
+
+    if (build > 22000) {
+      // After Windows 11
+      return true;
+    } else if (build === 22000) {
+      // Windows 11
+      return patch >= 1281;
+    } else {
+      // Windows 10
+      return patch >= 2311;
+    }
+  }
+
+  /**
+   * Check if the WSL store package is installed; returns a description of the
+   * error if it is not. If it is installed, returns an empty string.
+   */
+  async checkWSLFromStore(): Promise<string> {
+    try {
+      const pkg = 'MicrosoftCorporationII.WindowsSubsystemForLinux';
+      const command = `Get-AppxPackage -Name ${ pkg } | ConvertTo-Json`;
+      const { stdout } = await spawnFile(this.powerShellExecutable, [
+        '-NoLogo',
+        '-NoProfile',
+        '-NonInteractive',
+        '-WindowStyle', 'Hidden',
+        '-Command', command,
+      ], { stdio: ['ignore', 'pipe', console] });
+
+      if (!stdout) {
+        return 'WSL does not appear to be installed from the Microsoft Store.';
+      }
+      const result = JSON.parse(stdout || '{"missing": true}');
+
+      if (result.Status === 0) {
+        return '';
+      }
+      if (result.missing) {
+        return 'The Windows in-box WSL is installed instead of the Microsoft Store version.';
+      }
+      console.debug(`${ this.id }: ${ JSON.stringify(result, undefined, 2) }`);
+
+      return 'WSL was installed from the Microsoft Store, but it is not functioning correctly.';
+    } catch (ex) {
+      console.error(`Failed to check for WSL in store: ${ ex }`);
+
+      return 'There was an error checking if WSL was installed from the Microsoft Store.';
+    }
+  }
+
+  /**
+   * Run WSL and return stdout.
+   * @param args Command line for WSL
+   * @param failureMatch If given, ignore non-zero exit code if the output
+   *   contains this string.
+   */
+  async runWSL(args: string[], failureMatch?: string): Promise<string> {
+    try {
+      const result = await spawnFile(
+        path.join(this.system32, 'wsl.exe'), args,
+        { stdio: ['ignore', 'pipe', console], encoding: 'utf16le' });
+
+      return result.stdout;
+    } catch (ex: any) {
+      if (failureMatch && 'stdout' in ex) {
+        const stdout: string = ex.stdout;
+
+        if (stdout.includes(failureMatch)) {
+          return stdout;
+        }
+      }
+      throw ex;
+    }
+  }
+
+  _cachedWSLHelpText: string | undefined;
+  async getWSLHelpText(): Promise<string> {
+    if (this._cachedWSLHelpText) {
+      return this._cachedWSLHelpText;
+    }
+    // `wsl.exe --help` always exits with -1; that's okay as long as the output
+    // contains `wsl.exe` somewhere.
```

Comment:
is this guaranteed? 😄

Reply:
I'm filtering for `wsl.exe` because that's in the actual help output (make sense, it's part of the usage string)… I think?

I'm trying to pick something that works on non-English Windows :) Maybe I should just switch it to German and hope I can read enough of it to get things to work…

Reply:
Hmm, this is annoying.
`wsl --help` when WSL _isn't_ installed appears to be always in English. However, once it's installed, then the help output is localized.  (It looks like the in-box version has incomplete help translations, but the store version is more fully translated, at least for German. Can't actually read any of it though…)